colossalai/nn/_ops/linear.py

```diff
@@ -86,8 +87,63 @@ def colo_linear_imp(input_tensor: GeneralTensor,
     return ret_tensor
 
 
+def _new_colo_linear_imp(input_tensor: GeneralTensor,
+                         weight: GeneralTensor,
+                         bias: Optional[GeneralTensor] = None) -> 'ColoTensor':
+    """
+    A tentative function to compute the distributed linear layer with the latest sharding spec.
+    This function is subject to future change as the current sharding API is not stable.
+    """
+    # get mesh info
+    device_mesh = weight.sharding_spec.device_mesh
+    pg = weight.get_process_group()
+
+    # handle column and row parallel linear
+    # by reusing the implementation above
+    if str(weight.sharding_spec.sharding_sequence) == '[R, S1]':
+        assert bias is None or str(bias.sharding_spec.sharding_sequence) == '[R]'
+        # the output is a replica
+        output_partition_dict = {}
+
+        # compute the matmul
+        partial_output = F.linear(input_tensor, weight)
+        ret_tensor = reduce_input(partial_output, pg)
+
+        # add bias
+        if bias is not None:
+            ret_tensor += bias
+    elif str(weight.sharding_spec.sharding_sequence) == '[S1, R]':
```

Comment:
[S0, R] is a valid sharding sequence for weight, like Matmul([R, R], [R, S0]) -> [R, S0]